From: jiyun_ha <jiyun_ha@tmax.co.kr>
Date: Wed, 29 Jun 2022 16:44:39 +0900
Subject: [feat] dash-to-panel : change tray volume menu icon

 - added new icon to volume menu that change the shape depending on the volume

[IMS] N/A
---
 extensions/dash-to-panel/tray.js | 37 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 36 insertions(+), 1 deletion(-)

diff --git a/extensions/dash-to-panel/tray.js b/extensions/dash-to-panel/tray.js
index 497712a..d270b9f 100644
--- a/extensions/dash-to-panel/tray.js
+++ b/extensions/dash-to-panel/tray.js
@@ -1,5 +1,5 @@
 /* exported Tray */
-const { Clutter, Gio } = imports.gi;
+const { Clutter, Gio, St } = imports.gi;
 
 const Me = imports.misc.extensionUtils.getCurrentExtension();
 const Utils = Me.imports.utils;
@@ -69,6 +69,8 @@ class Tray {
             }
         );
 
+        this._changeOutputVolumeIcon();
+
         let indicators = this._indicators.get_children();
 
         this._connectedSignals = [];
@@ -88,6 +90,8 @@ class Tray {
 
         this._indicators.get_last_child().visible = true;
 
+        this._revertOutputVolumeIcon();
+
         let power = this._originalTray._power;
         power.visible = true;
 
@@ -158,4 +162,35 @@ class Tray {
 
         return Clutter.EVENT_STOP;
     }
+
+    _changeOutputVolumeIcon() {
+        let volume = this._originalTray._volume;
+        let output = volume._volumeMenu._output;
+
+        let icon = new St.Icon({ style_class: 'popup-menu-icon' });
+        let originalIcon = output._icon;
+
+        output.item.add(icon);
+        output.item.set_child_below_sibling(icon, originalIcon);
+
+        icon.visible = true;
+        originalIcon.visible = false;
+
+        icon.icon_name = output.getIcon();
+        let id = output.connect('stream-updated', () => {
+            if (icon.icon_name !== output.getIcon())
+                icon.icon_name = output.getIcon();
+
+        });
+        icon.connect('destroy', () => {
+            output.disconnect(id);
+        });
+
+        this._volumeMenuIcon = icon;
+    }
+
+    _revertOutputVolumeIcon() {
+        this._originalTray._volume._volumeMenu._output._icon.visible = true;
+        this._volumeMenuIcon.destroy();
+    }
 }
